### PR TITLE
refactor: replace handwritten parser with library

### LINKED ISSUE #2975: REF: Remove the handwritten type parser
This parser is pretty overkill, and probably broken in various ways. We should try farming this out to another library like https://github.com/sighingnow/parsec.py.

### PR DESCRIPTION
**Pull request by cpcloud**
This PR replaces our current handwritten parser with a parser combinator
produced by the [`parsy`](https://parsy.readthedocs.io/) library.

Closes #2975.

Depends on https://github.com/conda-forge/staged-recipes/pull/16171

### CODE REVIEW

**Comment by jreback** [comment]
can you add a release note (indicating now depending on this) & that *some* parsing error result types might change. Also does this need to be in the official requirements somewhere? (setup.py) / readme

**Comment by cpcloud** [comment]
> can you add a release note (indicating now depending on this) & that _some_ parsing error result types might change. Also does this need to be in the official requirements somewhere? (setup.py) / readme

Release note added.

`requirements.txt` and `environment.yml` have both been updated, though see the PR description, as I'm waiting on a feedstock to be created https://github.com/conda-forge/staged-recipes/pull/16171.

File: docs/source/release/index.rst [hunk 1]
```diff
@@ -12,6 +12,7 @@ Release Notes
    These release notes are for versions of ibis **1.0 and later**. Release
    notes for pre-1.0 versions of ibis can be found at :doc:`release-pre-1.0`
 
+* :support:`2977` Remove handwritten type parser; parsing errors that were previously `IbisTypeError` are now `parsy.ParseError`.
```
Comment (jreback) [inline docs/source/release/index.rst:15]: also say that `parsy` is not a requirement

Comment (jreback) [reply docs/source/release/index.rst:15]: IIRC these are listed in setup.py as well?

Comment (cpcloud) [reply docs/source/release/index.rst:15]: `parsy` is a requirement though, is that what you meant?

**Comment by cpcloud** [comment]
@jreback I think this is good to go, the previous build passed. I think we can merge when the current build is green :)

**Review by jreback — APPROVED** [review]
nice!

**Pull request merged**